packages/flutter/lib/src/widgets/routes.dart

```diff
@@ -2486,3 +2508,51 @@ abstract class PopEntry<T> {
     return 'PopEntry canPop: ${canPopNotifier.value}, onPopInvoked: $onPopInvokedWithResult';
   }
 }
+
+/// Mixin for a route that can provide a delegated secondary transition to the
+/// outgoing route.
+mixin FlexibleTransitionRouteMixin<T> on ModalRoute<T> {
+  /// The delegated transition provided to the previous route.
+  DelegatedTransitionBuilder? get delegatedTransition;
+
+  /// The delegated transition received from an incoming route.
+  DelegatedTransitionBuilder? receivedTransition;
+
+  @override
+  DelegatedTransitionBuilder? get nextRouteTransition {
+    if (receivedTransition != null) {
+      return receivedTransition;
+    } else {
+      return null;
+    }
+  }
+
+  @override
+  bool didPop(T? result) {
+    final bool popResult = super.didPop(result);
+    if (popResult && receivedTransition != null) {
+      receivedTransition = null;
```

Comment:
Discussed in person why this was needed: The case is when pushing two routes and then popping them in succession. Call them Route1, 2 and 3.

When Route3 is popped, Route2 is using its delegatedTransition. When that transition finishes and Route3 is long gone, Route2 still has Route3's delegatedTransition in its widget tree! And not its own usual transitions provided by something like [PageTransitionsBuilder.buildTransitions](https://main-api.flutter.dev/flutter/material/PageTransitionsBuilder/buildTransitions.html). When Route2 is then popped, this didPop method is called and its receivedTransition is set to null so that it can go back to using its own original transition. But, that also causes the problem given in my comment a few lines above.

Reply:
@MitchellGoodwin That makes me think of another possible bug here. Same setup as before: push Route2 and Route3, then pop Route3. Now instead of popping Route2, push a new Route4, which does not have a delegatedTransition. If I'm understanding correctly, Route2 will not animate out correctly.

We should add a test for that too.

Reply:
Sorry for the late response. That last scenario will actually currently work. If Route4 does not have a delegate transition, then Route2 will set its `reveivedTransition` to null, so it will have it's original transitions. Whether it plays that transition or not when Route4 is pushed on top is up to `canTransitionTo` and `canTransitionFrom`, which is current behavior before this PR.